### ISSUE TITLE
[3.3.5][6.x] Core/Creature: Missing Z Position update in Combat

### ISSUE THREAD

**Issue by sirikfoll** [TrinityCore/TrinityCore#16549]
Example:
The boss gets out of melee reach, but for some unknown reason, he do not move closer to the player
Does not seem to be an "out of LoS" problem, because if used something like `me->IsWithinLOS(me-GetVictim())` , the return value will be true.
Guess the Z position value is not being handled well somewhere

Video: https://www.youtube.com/watch?v=i1e_UqyJTAs

vmaps and mmaps are enabled.

hash b4b43d03b3fb
branch 3.3.5

**Comment by P-Kito** [comment]
Oh, probably since https://github.com/TrinityCore/TrinityCore/commit/89d8d50b2e1185a14896bd8b8b6679da578d4ed4

**Comment by Keader** [comment]
@P-Kito nop, it happened before that.

**Comment by Rushor** [comment]
i think this can be reproduced with all creatures. (@Keader, yep also happens on a 2 years old rev.)

**Comment by Shauren** [comment]
It looks like creature is in range of player just fine (2d distance) but attack range is checked in 3d so it fails like that

**Comment by P-Kito** [comment]
Anyone who is interested in checking the math behind this, here you can find it: 

bool Unit::IsWithinMeleeRange(const Unit* obj, float dist)
bool Unit::IsWithinCombatRange(const Unit* obj, float dist2compare)

**Comment by Demonid** [comment]
Didn't someone say that if the boss cant reach the player, it must go into evade mode ? Even if that not the issue over here, it still a nice feature that should be added.

**Comment by P-Kito** [comment]
That doesn't fix the issue, the boss will then randomly evade if we don't fix this problem first.

**Comment by chaodhib** [comment]
The problem lies in `TargetedMovementGeneratorMedium<T, D>::_setTargetLocation`. Except for the fact that this method needs a rewritte because it's a mess, it uses `i_target->GetContactPoint(owner, x, y, z);` which works in 2D: 
```
GetNearPoint2D(x, y, distance2d+searcher_size, absAngle);
z = GetPositionZ();
...
```

I'll see what I can do.

**Comment by chaodhib** [comment]
Is it even possible to walk up these stairs when the Lich King is engaged in retail? Cause handling correctly this type of situation (mob with a big combat reach + high slope) will complicate the code, without adding much benefit for most of the time.

**Comment by chaodhib** [comment]
> I stood with my back to the stairs leading up to the throne, which act as an invisible barrier that protects against knockbacks
source: http://sonofalich.com/the-road-to-arthas-ruminations-reflections-and-strategy/

If I understand correctly, the staircase should be inaccessible while the Lich King is alive.

**Comment by Nawuko** [comment]
On retail the stairs are accessible during the fight, you can tank lichking up there and of course the melee is working as intended. The "invisible barrier" are the stairs itself. Cause if you get knocked back on retail and have terrain with a certain slope behind you the knock back is mitigated (cause the knock back stops due to impact with the stairs).

I can provide a video from current retail if needed.

**Comment by Keader** [comment]
Any news @chaodhib ?

**Comment by Keader** [comment]
In Maexxna's fight, this bug happens in alot of spots.
Need some priority here, very abusive bug.
And unlike Lich King, Z Boundary cant help here :(

**Comment by sirikfoll** [comment]
No longer valid, at least the problem with LK seems to be fixed by 2a84562dc85516f432bb1e5de9add23c28b26ce4
